### DISCUSSION TITLE
Scalability issues of expandable connectors in large systems

### DISCUSSION THREAD

**Discussion by casella** [OpenModelica/OpenModelica#12160] [Ideas]
Expandable connectors are ideal to mimick the concept of data buses in complex systems. The bus connections carry around all the input/output variables which are needed for monitoring and control in one "wire", then every sub-system or component reads or writes the variables it needs or generates using "labels" (i.e., component names on the expandable connector).

Consider a large system with, say 2000 variables on the bus and 500 bus connectors. Such number are not uncommon in real-life systems, that can actually have much larger values. By applying the [rules for expandable connectors](https://specification.modelica.org/maint/3.6/connectors-and-connections.html#expandable-connectors) in the MLS, each of these 500 bus connectors will contain the whole set of connector variables, with a total of 1M variables and equations, even though in most cases only a handful of the 2000 variables are actually needed: for example, if I connect a temperature sensor to the bus, I only need to add the equation `plant.unit34.temperatureSensor.T = plant.unit34.bus.T0034` to the system DAE, while all the other 1999 variables in `plant.unit34.bus` are totally irrelevant.

It is true that eventually those irrelevant variables will be dealt with by alias elimination, but my argument is: what is the point of generating O(N^2) totally useless expandable connector variables and equations, only to eliminate them afterwards?

Of course, if the number of bus variables and bus connectors is less than a few dozens, this is no big deal, as you get a few hundreds of aliases which are easily dealt with. But as you scale up the size, this quickly becomes a major, and totally unnecessary, burden.

I think this should be discussed in the context of OpenSCALING. @perost, @adrpo, @kabdelhak, @phannebohm, what do you think?

**Comment by gwr-de** [comment]
Maybe I have not clearly understood the issue here, but efficient use of connectors and connect-equations are dear to me as well (e.g., #12210). The use case for expandable connectors to me seems to be exactly related to the observed complexity of O(N^2). If I have 500 subsystems in a model (`m.s1` to `m.s500`) that exchange 2000 variables with each other then we might in the worst case need to write 2000 connect-equations, while we only need to write 500 connect equations when we are using an expandable connector `m.ec` and then connect each subsystem with that connector, e.g., `connect( m.s1.ec, m.ec),  ...`. In this thought example, I am neglecting the connections inside the subsystems, i.e., between inside and outside connectors, as these would hold with or without `expandable connector` use. (In other words, O(N^2) is reduced to O(N) with regard to connect-equations...)

Since we are using pass by reference—not value—can you explain where exactly there are inefficiencies?

**Reply by casella** [reply]
@gwr69 it is much worse than what you write. Say that each model from `m.s1` to `m.s1000` write one output `var_j` to the expandable connector `m.sxx.c` and `m.s0` reads all of them through `m.s0.c`. All things considered, 1000 signals are exchanged.

According to the current expandable connector specification, each connector `m.sxx.c` will contain the entire set of outputs `var_1...var_1000`, i.e., 1000 variables, because all expandable connectors in the connection sets, both inside and outside, contain all the variables. That gives 1 million variables in this model. Do you see the problem now?

**Reply by gwr-de** [reply]
@casella I am certainly talking way out of my depth (disclaimer: I am not a computer scientist), but to me an `expandable connector` is something like a "dynamic data type" and that to me means I need to store 1000 `connect-equations` to generate "at front end parsing run time (?)" all the information I need to build the "connection set"—not 1000 x 1000 (static) variables. Obviously, I am missing something on the implementation side?

**Reply by casella** [reply]
> (disclaimer: I am not a computer scientist)

Me neither 😄 

If you check the [MLS rules](https://specification.modelica.org/maint/3.6/connectors-and-connections.html#expandable-connectors) for expandable connectors, you read:

> When two expandable connectors are connected, each is augmented with the variables that are only declared in the other expandable connector (the new variables are neither input nor output). This is repeated until all connected expandable connector instances have matching variables

As I understand, this means that eventually each and every expandable connectors in a connection will contain the same variables. In the case of the 1000 connections mentioned in my previous post, each connector will eventually contain all the 1000 variables `var_1...var_1000`. Thus, 1000 connectors each containing 1000 variables means 1 M variables in the flat model.

@perost, are we on the same page, or do I miss something?

**Reply by gwr-de** [reply]
Yes, I am aware of MLS. But the specification does only describe _required behavior_—not how to implement it.

**Reply by casella** [reply]
Sorry, I'm not sure I understand what you mean. If the semantics of the expandable connectors explicitly calls for each of them to have 1000 variables and you have 1000 of them, then your flat model must contain 1 M variables:
- `m.s1.c.var_1`
- `m.s1.c.var_2`
- ...
-` m.s1.c_var_1000`
- `m.s2.c.var_1`
- `m.s2.c.var_2`
- ...
- `m.s1000.c.var_1000`
 
This doesn't seem to me an implementation issue. Of course you may generate them very efficiently, but they have to be there.

**Reply by casella** [reply]
Consider this MWE
```modelica
package TestExpandableConnectors4
   
  expandable connector C
  end C;

  model M1
  Modelica.Blocks.Sources.RealExpression exp(y = 0);
    C c;
  equation
    connect(exp.y, c.y1);
  end M1;

  model M2
  Modelica.Blocks.Sources.RealExpression exp(y = 0);
    C c;
  equation
    connect(exp.y, c.y2);
  end M2;

  model M3
  Modelica.Blocks.Sources.RealExpression exp(y = 0);
    C c;
  equation
    connect(exp.y, c.y3);
  end M3;

  model S
    M1 m1;
    M2 m2;
    M3 m3;
    C c;
  equation
    connect(m1.c, c);
    connect(m2.c, c);
    connect(m3.c, c);
  end S;
end TestExpandableConnectors4;
```
The flattened S model contains
```modelica
  Real m1.c.y1;
  Real m1.c.y2;
  Real m1.c.y3;
  Real m2.c.y1;
  Real m2.c.y2;
  Real m2.c.y3;
  Real m3.c.y1;
  Real m3.c.y2;
  Real m3.c.y3;
```
and of course it must contain all the equations to make y1 equal in all connectors, y2 equal in all connectors, etc:
```modelica
  m3.c.y1 = m2.c.y1;
  m3.c.y1 = m1.c.y1;
  m3.c.y2 = m2.c.y2;
  m3.c.y2 = m1.c.y2;
  m3.c.y3 = m2.c.y3;
  m3.c.y3 = m1.c.y3;
```
It is obvious that the size of this model grows as O(N^2) as you add more instances of M to it, connecting each of them to a different variable in the expandable connector.

**Reply by casella** [reply]
The point of my proposal is that there is actually no need to instantiate the unused variables inside each connector, and the corresponding equations. For example `m1.c.y2` and `m1.c.y3` are totally useless, so they could be skipped from the instantiation process, as well as their corresponding equations.
